pkg/toolsets/toolsets.go

```diff
@@ -30,288 +33,719 @@ func NewToolsetDoesNotExistError(name string) *ToolsetDoesNotExistError {
+// The context can be used to extract actor/user information for flag evaluation.
+// Returns (enabled, error). If error occurs, the caller should log and treat as false.
+type FeatureFlagChecker func(ctx context.Context, flagName string) (bool, error)
+
+// NewRegistry creates a new empty Registry.
+// Use SetTools, SetResources, SetPrompts to populate it.
+func NewRegistry() *Registry {
+	return &Registry{
+		deprecatedAliases: make(map[string]string),
 	}
-	return nil
 }
 
-func (t *Toolset) GetAvailableTools() []ServerTool {
-	if t.readOnly {
-		return t.readTools
-	}
-	return append(t.readTools, t.writeTools...)
+// SetTools sets the tools for this group. Returns self for chaining.
+func (r *Registry) SetTools(tools []ServerTool) *Registry {
+	r.tools = tools
+	return r
 }
 
-func (t *Toolset) RegisterTools(s *mcp.Server) {
-	if !t.Enabled {
-		return
+// SetResources sets the resource templates for this group. Returns self for chaining.
+func (r *Registry) SetResources(resources []ServerResourceTemplate) *Registry {
+	r.resourceTemplates = resources
+	return r
+}
+
+// SetPrompts sets the prompts for this group. Returns self for chaining.
+func (r *Registry) SetPrompts(prompts []ServerPrompt) *Registry {
+	r.prompts = prompts
+	return r
+}
+
+// copy creates a shallow copy of the Registry for immutable operations.
+func (r *Registry) copy() *Registry {
+	newTG := &Registry{
+		tools:             r.tools, // slices are shared (immutable)
+		resourceTemplates: r.resourceTemplates,
+		prompts:           r.prompts,
+		deprecatedAliases: r.deprecatedAliases,
+		readOnly:          r.readOnly,
+		featureChecker:    r.featureChecker,
 	}
-	for i := range t.readTools {
-		t.readTools[i].RegisterFunc(s, t.deps)
+
+	// Copy maps if they exist
+	if r.enabledToolsets != nil {
+		newTG.enabledToolsets = make(map[ToolsetID]bool, len(r.enabledToolsets))
+		for k, v := range r.enabledToolsets {
+			newTG.enabledToolsets[k] = v
+		}
 	}
-	if !t.readOnly {
-		for i := range t.writeTools {
-			t.writeTools[i].RegisterFunc(s, t.deps)
+	if r.additionalTools != nil {
+		newTG.additionalTools = make(map[string]bool, len(r.additionalTools))
+		for k, v := range r.additionalTools {
+			newTG.additionalTools[k] = v
 		}
 	}
-}
+	newTG.unrecognizedToolsets = r.unrecognizedToolsets
 
-// SetDependencies sets the dependencies for this toolset's tool handlers.
-// The deps parameter is typed as `any` to avoid circular dependencies between packages.
-func (t *Toolset) SetDependencies(deps any) *Toolset {
-	t.deps = deps
-	return t
+	return newTG
 }
 
-func (t *Toolset) AddResourceTemplates(templates ...ServerResourceTemplate) *Toolset {
-	t.resourceTemplates = append(t.resourceTemplates, templates...)
-	return t
+// WithReadOnly returns a new Registry with read-only mode set.
+// When true, write tools are filtered out from Available* methods.
+func (r *Registry) WithReadOnly(readOnly bool) *Registry {
+	newTG := r.copy()
+	newTG.readOnly = readOnly
```

Comment:
Looking at the example from PR description the intended use of this function as as a builder pattern:
```
filtered := tsg.
    WithReadOnly(true).
    WithToolsets([]string{"repos", "issues"}).
    WithTools([]string{"get_me"}).  // additive, bypasses toolset filter
    WithFeatureChecker(checker)  
```
In current implementation every builder function is unnecessarily copying the registry. Each `Registry.copy()` is iterating over tool slices to copy them too.
Maybe we could introduce a struct like `RegistryBuilder`, then functions as `WithReadOnly` will just set one field in the builder. Once the user is done with constructing a new object she would call `build()` on the builder instance, which returns the copied registry object with some fields overridden.